velox/functions/prestosql/aggregates/AverageAggregate.cpp

```diff
@@ -279,19 +309,98 @@ class AverageAggregate : public exec::Aggregate {
+                                 is_same_v<TAccumulator, int128_t>) {
+          // Need to perform integer division here as the intermediate sums may
+          // overflow decimal limits.
+          int128_t result = 0;
+          DecimalUtil::divideWithRoundUp<int128_t, int128_t, int64_t>(
+              result, sumCount->sum, sumCount->count, 0, 0);
+          rawValues[i] = TResult(result);
+        }
       }
     }
   }
 
   DecodedVector decodedRaw_;
   DecodedVector decodedPartial_;
-};
+}; // class AverageAggregate
+
+/// Override 'accumulatorAlignmentSize' for UnscaledLongDecimal values as it
+/// uses int128_t type. Some CPUs don't support misaligned access to int128_t
+/// type.
+template <>
+inline int32_t
+AverageAggregate<UnscaledShortDecimal, int128_t, UnscaledShortDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+template <>
+inline int32_t
+AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledLongDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+// This overload required for final aggregator after partially aggregating a
+// vector short decimals.
+template <>
+inline int32_t
+AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledShortDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+template <>
+template <bool tableHasNulls>
+void AverageAggregate<UnscaledShortDecimal, int128_t, UnscaledShortDecimal>::
+    updateNonNullValue(char* group, UnscaledShortDecimal value) {
+  if constexpr (tableHasNulls) {
+    exec::Aggregate::clearNull(group);
+  }
+  accumulator(group)->sum =
+      checkedPlus<int128_t>(accumulator(group)->sum, value.unscaledValue());
+  accumulator(group)->count += 1;
+}
+
+template <>
+template <bool tableHasNulls>
+void AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledLongDecimal>::
+    updateNonNullValue(char* group, UnscaledLongDecimal value) {
+  if constexpr (tableHasNulls) {
+    exec::Aggregate::clearNull(group);
+  }
+  accumulator(group)->sum =
+      checkedPlus<int128_t>(accumulator(group)->sum, value.unscaledValue());
+  accumulator(group)->count += 1;
+}
+
+template <>
+template <bool tableHasNulls>
+void AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledShortDecimal>::
+    updateNonNullValue(char* group, UnscaledLongDecimal value) {
+  if constexpr (tableHasNulls) {
+    exec::Aggregate::clearNull(group);
+  }
+  accumulator(group)->sum =
+      checkedPlus<int128_t>(accumulator(group)->sum, value.unscaledValue());
+  accumulator(group)->count += 1;
+}
 
 void checkSumCountRowType(TypePtr type, const std::string& errorMessage) {
-  VELOX_CHECK_EQ(type->kind(), TypeKind::ROW, "{}", errorMessage);
-  VELOX_CHECK_EQ(
-      type->childAt(0)->kind(), TypeKind::DOUBLE, "{}", errorMessage);
+  VELOX_CHECK(
+      type->kind() == TypeKind::ROW || type->kind() == TypeKind::LONG_DECIMAL,
```

Comment:
Why do we need only the LONG_DECIMAL check in `checkSumCountRowType`
Can you add a comment?

Reply:
can you address this comment?

Reply:
This is a mistake, the check must be split. Fixed it.